nixos/modules/services/security/crowdsec.nix

```diff
@@ -0,0 +1,841 @@
+          console = mkOption {
+            type = types.submodule {
+              options = {
+                token = mkOption {
+                  type = types.nullOr types.str;
+                  default = null;
+                  example = "abcde";
+                  description = ''
+                    The console token to enroll to the web console.
+                  '';
+                };
+                tokenFile = mkOption {
+                  type = types.nullOr types.path;
+                  example = "/run/crowdsec/console_token.yaml";
+                  description = ''
+                    The credential file to use. This is preferred instead of putting secrets in the Nix store.
+                  '';
+                  default = null;
+                };
+                configuration = mkOption {
+                  type = format.type;
+                  default = {
+                    share_manual_decisions = false;
+                    share_custom = false;
+                    share_tainted = false;
+                    share_context = false;
+                  };
+                  description = ''
+                    Attributes inside the console.yaml file.
+                  '';
+                };
+              };
+            };
+            description = ''
+              Console Configuration attributes
+            '';
+            default = { };
+          };
+        };
+      };
+    };
+  };
+  config =
+    let
+      cfg = config.services.crowdsec;
+      configFile = format.generate "crowdsec.yaml" cfg.settings.general;
+      lapiFile =
+        if cfg.settings.lapi.credentialsFile != null then
+          cfg.settings.lapi.credentialsFile
+        else
+          format.generate "local_api_credentials.yaml" cfg.settings.lapi.credentials;
+      capiFile =
+        if cfg.settings.capi.credentialsFile != null then
+          cfg.settings.capi.credentialsFile
+        else
+          (
+            if cfg.settings.capi.credentials != null then
+              format.generate "central_api_credentials.yaml" cfg.settings.capi.credentials
+            else
+              null
+          );
+
+      tokenFile =
+        if cfg.settings.console.tokenFile != null then
+          cfg.settings.console.tokenFile
+        else
+          (
+            if cfg.settings.console.token != null then
+              pkgs.writeText "console_token.txt" cfg.settings.console.token
+            else
+              null
+          );
+      simulationFile = format.generate "simulation.yaml" cfg.settings.simulation;
+      consoleFile = format.generate "console.yaml" cfg.settings.console.configuration;
+      patternsDir = pkgs.buildPackages.symlinkJoin {
+        name = "crowdsec-patterns";
+        paths = [
+          cfg.localConfig.patterns
+          (lib.attrsets.getOutput "patterns" pkg)
+        ];
+      };
+
+      pkg = cfg.package;
+
+      cscli = pkgs.writeScriptBin "cscli" ''
+        #!${pkgs.runtimeShell}
```

Comment:
Besides the pointless double runtimeshell (`writeScriptBin` already adds that), the user of this script ends up wrong, causing permission issues if you don't use it carefully. In my module I took the approach that the `nextcloud` module takes:

```bash
export PATH="$PATH:${cfg.package}/bin/"

sudo=exec
if [ "$USER" != "crowdsec" ]; then
    sudo='exec /run/wrappers/bin/sudo -u crowdsec'
fi

$sudo ${cfg.package}/bin/cscli "$@"
```

That said, I'm not happy with that either, due to the reliance on sudo, and especially using the wrapper like this, but given other services do the same I think it's a lesser evil.

Reply:
`"$USER" != "crowdsec"` and `-u crowdsec` would also need to pull from the user set in `settings.crowdsec.user` no?

Reply:
Yep, replacing `crowdsec` with `${cfg.user}` should do the trick. If we don't place the configuration file in `/etc` a `-c ${cfg.settings}` would also be appropriate,